src/compose/ComposeBox.js

```diff
@@ -393,17 +401,26 @@ class ComposeBoxInner extends PureComponent<Props, State> {
     if (isStreamNarrow(narrow) || (isTopicNarrow(narrow) && isEditing)) {
       const streamName = streamNameOfNarrow(narrow);
       const topic = this.state.topic.trim();
-      return topicNarrow(streamName, topic || '(no topic)');
+      return topicNarrow(streamName, topic || apiConstants.NO_TOPIC_TOPIC);
     }
     invariant(isConversationNarrow(narrow), 'destination narrow must be conversation');
     return narrow;
   };
 
   handleSend = () => {
-    const { dispatch } = this.props;
+    const { dispatch, mandatoryTopics, _ } = this.props;
     const { message } = this.state;
     const destinationNarrow = this.getDestinationNarrow();
 
+    if (
+      isTopicNarrow(destinationNarrow)
+      && topicOfNarrow(destinationNarrow) === apiConstants.NO_TOPIC_TOPIC
+      && mandatoryTopics
+    ) {
+      showErrorAlert(_('Message not sent'), _('Please specify a topic.'));
+      return;
+    }
+
     this.props.onSend(message, destinationNarrow);
```

Comment:
So, this code actually runs not only when you send a message you've just composed in the normal way, but also when you hit save/submit on a message you've just gone and edited. (Maybe that's a sign we should rename this method, and also the `onSend` callback -- more like `handleSubmit` / `onSubmit`.)

How does this behave in the editing case? Seems like it'll give a confusing error message, if nothing else. What's even the desired semantics in that case -- if a message is already to `(no topic)`, and then the setting is enabled, and then you go to edit it, can you edit it without changing its topic?

… I kind of don't want to let those questions block this PR, because that's a pretty rare case (editing is already uncommon, and editing with a target topic of `(no topic)` when that's forbidden is surely much more so), and because they seem like questions that may lead to a discussion of what the web app's behavior is and should be. Better to get the main issue fixed here, and leave that for another time.

But let's leave at least a TODO comment, like `// TODO how should this behave in the isEditing case?`. That way this at least doesn't much add to the confusingness of this code with respect to its reuse for editing vs. composing: the reader sees it's specifically about sending a message, but the comment warns that doesn't mean the function is only for the sending case.